File: buildSrc/src/main/java/io/opentelemetry/instrumentation/gradle/muzzle/MuzzleDirective.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.gradle.muzzle;

import java.util.Collections;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import org.gradle.api.provider.ListProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.provider.SetProperty;

public abstract class MuzzleDirective {

  private static final Pattern NORMALIZE_NAME_SLUG = Pattern.compile("[^a-zA-Z0-9]+");

  public MuzzleDirective() {
    getName().convention("");
    getSkipVersions().convention(Collections.emptySet());
    getAdditionalDependencies().convention(Collections.emptyList());
    getExcludedDependencies().convention(Collections.emptyList());
    getAssertPass().convention(false);
    getAssertInverse().convention(false);
    getCoreJdk().convention(false);
  }

  public abstract Property<String> getName();

  public abstract Property<String> getGroup();

  public abstract Property<String> getModule();

  public abstract Property<String> getVersions();

  public abstract SetProperty<String> getSkipVersions();

  public abstract ListProperty<String> getAdditionalDependencies();

  public abstract ListProperty<String> getExcludedDependencies();

  public abstract Property<Boolean> getAssertPass();

  public abstract Property<Boolean> getAssertInverse();

  public abstract Property<Boolean> getCoreJdk();

  public void coreJdk() {
    getCoreJdk().set(true);
  }

  /**
   * Adds extra dependencies to the current muzzle test.
   *
   * @param compileString An extra dependency in the gradle canonical form:
   *     '<group_id>:<artifact_id>:<version_id>'.
   */
  public void extraDependency(String compileString) {
    getAdditionalDependencies().add(compileString);
  }

  /**
   * Adds transitive dependencies to exclude from the current muzzle test.
   *
   * @param excludeString A dependency in the gradle canonical form: '<group_id>:<artifact_id>'
   */
  public void excludeDependency(String excludeString) {
    getExcludedDependencies().add(excludeString);
  }

  public void skip(String... version) {
    getSkipVersions().addAll(version);
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    if (getCoreJdk().getOrElse(false)) {
      if (getAssertPass().getOrElse(false)) {
        sb.append("Pass");
      } else {
        sb.append("Fail");
      }
      sb.append("-core-jdk");
    } else {
      if (getAssertPass().getOrElse(false)) {
        sb.append("pass");
      } else {
        sb.append("fail");
      }
      sb.append(getGroup().get())
          .append(':')
          .append(getModule().get())
          .append(':')
          .append(getVersions().get());
    }
    return sb.toString();
  }

  String getNameSlug() {
    return NORMALIZE_NAME_SLUG.matcher(getName().get().trim()).replaceAll("-");
  }

  Set<String> getNormalizedSkipVersions() {
    return getSkipVersions().getOrElse(Collections.emptySet()).stream()
        .map(String::toLowerCase)
        .collect(Collectors.toSet());
  }
}

File: gradle-plugins/src/main/kotlin/io/opentelemetry/javaagent/muzzle/MuzzleDirective.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.javaagent.muzzle

import org.gradle.api.provider.ListProperty
import org.gradle.api.provider.Property
import org.gradle.api.provider.SetProperty
import java.util.stream.Collectors

abstract class MuzzleDirective {

  abstract val name: Property<String>
  abstract val group: Property<String>
  abstract val module: Property<String>
  abstract val classifier: Property<String>
  abstract val versions: Property<String>
  abstract val skipVersions: SetProperty<String>
  abstract val additionalDependencies: ListProperty<Any>
  abstract val excludedDependencies: ListProperty<String>
  abstract val excludedInstrumentationNames: SetProperty<String>
  abstract val assertPass: Property<Boolean>
  abstract val assertInverse: Property<Boolean>
  internal abstract val coreJdk: Property<Boolean> // use coreJdk() function below to enable

  init {
    name.convention("")
    classifier.convention("")
    skipVersions.convention(emptySet())
    additionalDependencies.convention(listOf())
    excludedDependencies.convention(listOf())
    excludedInstrumentationNames.convention(listOf())
    assertPass.convention(false)
    assertInverse.convention(false)
    coreJdk.convention(false)
  }

  fun coreJdk() {
    coreJdk.set(true)
  }

  /**
   * Adds extra dependencies to the current muzzle test.
   *
   * @param dependencyNotation An extra dependency in the gradle canonical form:
   * '<group_id>:<artifact_id>:<version_id>' or a project dependency project(':some-project').
   */
  fun extraDependency(dependencyNotation: Any) {
    additionalDependencies.add(dependencyNotation)
  }

  /**
   * Adds transitive dependencies to exclude from the current muzzle test.
   *
   * @param excludeString A dependency in the gradle canonical form: '<group_id>:<artifact_id>'
   */
  fun excludeDependency(excludeString: String?) {
    excludedDependencies.add(excludeString!!)
  }

  /**
   * Excludes an instrumentation module from the current muzzle test.
   *
   * @param excludeString An instrumentation module class name to exclude
   */
  fun excludeInstrumentationName(excludeString: String) {
    excludedInstrumentationNames.add(excludeString)
  }

  fun skip(vararg version: String?) {
    skipVersions.addAll(*version)
  }

  internal val nameSlug: String
    get() = NORMALIZE_NAME_SLUG.replace(name.get().trim(), "-")

  internal val normalizedSkipVersions: Set<String>
    get() = skipVersions.getOrElse(setOf()).stream()
      .map(String::toLowerCase)
      .collect(Collectors.toSet())

  override fun toString(): String {
    val sb = StringBuilder()
    if (coreJdk.getOrElse(false)) {
      if (assertPass.getOrElse(false)) {
        sb.append("Pass")
      } else {
        sb.append("Fail")
      }
      sb.append("-core-jdk")
    } else {
      if (assertPass.getOrElse(false)) {
        sb.append("pass")
      } else {
        sb.append("fail")
      }
      sb.append(group.get())
        .append(':')
        .append(module.get())
        .append(':')
        .append(versions.get())
      if (classifier.isPresent) {
        sb.append(':').append(classifier.get())
      }
    }
    return sb.toString()
  }

  companion object {
    private val NORMALIZE_NAME_SLUG = Regex("[^a-zA-Z0-9]+")
  }
}
